driver/gl/gl.go

```diff
@@ -172,113 +145,45 @@ func (c *glCanvas) newGlTextTexture(obj fyne.CanvasObject) uint32 {
-
-			w, h := int(icon.ViewBox.W), int(icon.ViewBox.H)
-			// this is used by our render code, so let's set it to the file aspect
-			img.PixelAspect = float32(w) / float32(h)
-			// if the image specifies it should be original size we need at least that many pixels on screen
-			if img.FillMode == canvas.ImageFillOriginal {
-				pixSize := fyne.NewSize(unscaleInt(c, w), unscaleInt(c, h))
-				img.SetMinSize(pixSize)
-			}
-
-			raw = image.NewRGBA(image.Rect(0, 0, width, height))
-			scanner := rasterx.NewScannerGV(w, h, raw, raw.Bounds())
-			raster := rasterx.NewDasher(width, height, scanner)
-
-			icon.Draw(raster, 1)
-		} else {
-			pixels, _, err := image.Decode(file)
-
-			if err != nil {
-				log.Println("image err", err)
-
-				return 0
-			}
-			origSize := pixels.Bounds().Size()
-			// this is used by our render code, so let's set it to the file aspect
-			img.PixelAspect = float32(origSize.X) / float32(origSize.Y)
-			// if the image specifies it should be original size we need at least that many pixels on screen
-			if img.FillMode == canvas.ImageFillOriginal {
-				pixSize := fyne.NewSize(unscaleInt(c, origSize.X), unscaleInt(c, origSize.Y))
-				img.SetMinSize(pixSize)
-			}
-
-			raw = image.NewRGBA(pixels.Bounds())
-			draw.Draw(raw, pixels.Bounds(), pixels, image.ZP, draw.Src)
-		}
-	} else if img.PixelColor != nil {
-		raw = image.NewRGBA(image.Rect(0, 0, width, height))
-		pixels := newPixelImage(img, c.Scale())
-
-		halfWidth := raw.Bounds().Size().X / 2
-		halfHeight := raw.Bounds().Size().Y / 2
+	if img.Raster != nil {
+		return c.img2Texture(img.Raster(width, height))
+	}
 
-		// use a WaitGroup so we don't return our image until it's complete
-		var wg sync.WaitGroup
-		wg.Add(4)
+	return c.img2Texture(image.NewRGBA(image.Rect(0, 0, 1, 1)))
 
-		go renderGlImagePortion(image.ZP, halfWidth, halfHeight, raw, pixels, &wg)
-		go renderGlImagePortion(image.Pt(0, halfHeight), halfWidth, height-halfHeight, raw, pixels, &wg)
-		go renderGlImagePortion(image.Pt(halfWidth, 0), width-halfWidth, halfHeight, raw, pixels, &wg)
-		go renderGlImagePortion(image.Pt(halfWidth, halfHeight), width-halfWidth, height-halfHeight, raw, pixels, &wg)
+}
 
-		wg.Wait()
-	} else {
-		raw = image.NewRGBA(image.Rect(0, 0, 1, 1))
+func (c *glCanvas) img2Texture(img image.Image) uint32 {
+	switch i := img.(type) {
+	case (*image.RGBA):
+		texture := newTexture()
+		gl.TexImage2D(gl.TEXTURE_2D, 0, gl.RGBA, int32(i.Rect.Size().X), int32(i.Rect.Size().Y),
+			0, gl.RGBA, gl.UNSIGNED_BYTE, gl.Ptr(i.Pix))
+		return texture
+	case (*image.Uniform):
+		texture := newTexture()
+		r, g, b, a := i.RGBA()
+		r8, g8, b8, a8 := uint8(r>>8), uint8(g>>8), uint8(b>>8), uint8(a>>8)
+		data := []uint8{r8, g8, b8, a8}
+		gl.TexImage2D(gl.TEXTURE_2D, 0, gl.RGBA, 1, 1, 0, gl.RGBA,
+			gl.UNSIGNED_BYTE, gl.Ptr(data))
+		return texture
+	default:
+		rgba := image.NewRGBA(image.Rect(0, 0, img.Bounds().Dx(), img.Bounds().Dy()))
+		return c.img2Texture(rgba)
```

Comment:
Before this return you are not drawing the old pixels into the new image. I think this line is missing:

`draw.Draw(rgba, rgba.Rect, img, image.ZP, draw.Over)`

Reply:
Thanks, this fixed the image rendering!